docs/pipelines.md

```diff
@@ -0,0 +1,87 @@
+# Pipelines
+
+So you've used these libraries to set up some pipelines. Great! But
+what's next? This will explain our common patterns and how to use them.
+
+## Stack pipelines
+
+For a stack repository, we typically create a few different jobs:
+
+* `branch-create`
+* `tag`
+* `publish`
+* `continuous-publish`
+
+Each of them has a different purpose.
+
+### What repository can I use as a reference?
+
+Try out the [sample-stack-wordpress repo](https://github.com/crossplane/sample-stack-wordpress).
+Take a look at the `Jenkinsfile.*` files.
+
+### What else is needed other than Jenkinsfiles?
+
+You'll also need to set up permissions and webhooks:
+
+* The group `crossplane/upbound-bot` needs `Write` permissions for the
+  repository so it can manage branches and tags during releases.
+* There should be a webhook set up so Jenkins can run a continuous
+  release when `master` is updated:
+  - Payload URL: `https://jenkinsci.upbound.io/github-webhook/`
+  - Content type: `application/x-www-form-urlencoded`
+  - Secret: (blank)
+  - SSL Verfication: `Enable SSL verification`
+  - Which events would you like to trigger this webhook?: `Just the push event`
+
+
+### I have a new repository, and I've added the appropriate Jenkinsfiles. How do I create the pipeline jobs?
+
+Typically we copy from an existing item. The general steps are:
+
+* Go to the [upbound-jenkins].
+* Log in
+* Create a new item in a folder matching the org name (typically
+  `crossplane`).
+* At the top, enter in the name of your repository, such as
+  `stack-my-app`.
+* At the bottom, in the `Copy from` field, enter in an existing stack
+  folder, such as `sample-stack-wordpress`.
+* Once the folder is created, go into each of the jobs that was created,
+  and update the Git url to point to your repository.
+
+### How do I cut a release?
+
+Cutting a release involves: (potentially) cutting a new release branch;
+tagging a commit for a release; building and publishing from the tagged
+commit. Using the jobs, the steps are:
+
+1. Using the job `branch-create`, create a release branch if one doesn't
+   exist. We use the convention `release-MAJOR.MINOR`. For example:
+   `release-0.1`. If a branch already exists for the release version
+   prefix, update the branch with the new changes instead of creating a
+   new one.
+2. Using the job `tag`, tag the commit you want to release, with the
+   release tag. The tag should be in the form `vMAJOR.MINOR.PATCH`. For
+   example: `v0.1.0`.
+3. Using the job `publish`, build and publish the tag you want to
```

Comment:
How is the entity that pushes to Docker Hub (or potentially to other registries) configured?

Reply:
We use some environment variables for logging into docker. It looks like I must've removed the `docker login` at some point and [never put it back](https://github.com/crossplane/cicd/blob/df0d5adf23d6d514df07d53803037765e80baee1/vars/runStackPublishPipeline.groovy#L36), so we'll want to add that again to the jobs that use docker, to make sure they're reliable. For breadcrumbs, here's a link to an [example of logging in](https://github.com/crossplane/provider-gcp/blob/8d284b7d453a00a1725bf75068856b066e74fd46/Jenkinsfile.promote#L23).

I assumed that the environment variables were being set by the `credentials` [line in the Jenkinsfile](https://github.com/crossplane/cicd/blob/df0d5adf23d6d514df07d53803037765e80baee1/vars/runStackPublishPipeline.groovy#L16), but I never verified that. They could be getting injected in some other way.

Reply:
Do you mean `DOCKER = credentials('dockerhub-upboundci')` is the line that handles the login? If so, could you add some comments regarding that or maybe do an explicit `docker login`?

Reply:
Not exactly; I believe that line injects the credentials we need, but that there should be a line that runs `docker login` explicitly, since that's what the other jobs do. I copied some working configurations originally and never inspected all of the details, so there are some gaps in my knowledge of how exactly the credentials become available.

Reply:
I think this is a pretty important detail when setting up a new pipeline for a repo. Would you mind investigating how it's working and update the setup section above?

Reply:
In this context, it is not important for setting up a new pipeline because we use a single set of credentials between our different repositories, and because our code reuse mechanism allows us to easily share the configuration to use those credentials. In other words, there are no steps required for a person who is setting up a new pipeline.

It would be helpful for people who want to use a different docker registry with different credentials, or if we had a more granular access control strategy, but that is out of scope for this draft of the instructions. We can always add that to a future document when we want to make it easier for people to do that.